### PR TITLE
Fix sonarcloud bugs of sut-api code

### PR DESCRIPTION
**Pull request by enesoezel**

### CODE REVIEW

File: src/main/java/de/retest/configuration/Configuration.java [hunk 1]
```diff
@@ -22,13 +22,9 @@ static boolean isLoaded() {
 		return instance != null && instance.retestWorkspace != null;
 	}
 
-	public static void ensureLoaded() {
+	public static synchronized void ensureLoaded() {
```
Comment (diba1013) [inline src/main/java/de/retest/configuration/Configuration.java]: I would rather prefer using a special lock with synchronize(lock) since using synchronized in a method locks on the object's instance. As a consequence if a other method uses the another synchronize on a method, which is completely unrelated to this method, we would synchronize without needing to.

File: src/main/java/de/retest/persistence/migration/XmlMigrator.java [hunk 2]
```diff
@@ -47,7 +47,11 @@ public static NamedBufferedInputStream tryToMigrate( final XmlVersionCheckResult
 			for ( int i = 0; i < transformers.size(); i++ ) {
 				final boolean notFinalInputStream = i < transformers.size() - 1;
 				final InputStream previousInputStream = finalInputStream;
-				finalInputStream = transformers.get( i ).transform( finalInputStream );
+				try {
+					finalInputStream = transformers.get( i ).transform( finalInputStream );
+				} catch ( final IOException e ) {
+					throw new RuntimeException( e );
```
Comment (diba1013) [inline src/main/java/de/retest/persistence/migration/XmlMigrator.java]: In the below try-catch we just log the error. Which method is preferred since we exit the execution here? In either case, I would log the error.

File: src/main/java/de/retest/persistence/migration/XmlTransformer.java [hunk 3]
```diff
@@ -57,19 +55,13 @@ public final InputStream transform( final InputStream inputStream ) {
...
 			eventReader.close();
 			eventWriter.flush();
 			eventWriter.close();
-			out.close();
 
 			return new LZ4BlockInputStream( new DeleteOnCloseFileInputStream( tmpFile ) );
-		} catch ( final XMLStreamException e ) {
-			throw new RuntimeException( e );
-		} catch ( final FactoryConfigurationError e ) {
+		} catch ( final XMLStreamException | FactoryConfigurationError | IOException e ) {
```
Comment (diba1013) [inline src/main/java/de/retest/persistence/migration/XmlTransformer.java]: Why do you catch IOException, if you specify in the method that it is allowed to throw?

File: src/main/java/de/retest/persistence/migration/XmlTransformer.java [hunk 4]
```diff
@@ -57,19 +55,13 @@ public final InputStream transform( final InputStream inputStream ) {
 				final XMLEvent nextEvent = eventReader.nextEvent();
 				convert( nextEvent, eventWriter );
 			}
-
 			eventReader.close();
 			eventWriter.flush();
 			eventWriter.close();
```
Comment (diba1013) [inline src/main/java/de/retest/persistence/migration/XmlTransformer.java:71]: It has nothing really to do with your PR, but could we use try-with-resources for the above close?

File: src/main/java/de/retest/ui/diff/IdentifyingAttributesDifferenceFinder.java [hunk 5]
```diff
@@ -24,12 +25,8 @@ public IdentifyingAttributesDifferenceFinder( final RecheckIgnore recheckIgnore
...
 	public IdentifyingAttributesDifference differenceFor( final IdentifyingAttributes expected,
 			final IdentifyingAttributes actual ) {
-		if ( expected == null ) {
-			throw new NullPointerException( "Expected cannot be null!" );
-		}
-		if ( actual == null ) {
-			throw new NullPointerException( "Actual cannot be null!" );
-		}
+		Objects.requireNonNull( expected );
+		Objects.requireNonNull( actual );
```
Comment (diba1013) [inline src/main/java/de/retest/ui/diff/IdentifyingAttributesDifferenceFinder.java]: We do not need checks here, since the variables are used below. Let the NPE be implicit.

File: src/main/java/de/retest/ui/diff/IdentifyingAttributesDifferenceFinder.java [hunk 6]
```diff
@@ -67,8 +64,14 @@ private static boolean pathDiffers( final IdentifyingAttributes expected, final
 
 	// We treat null == "". This is OK for visible attributeDifferences but less suitable for invisible ones.
 	private static boolean differs( final Object expected, final Object actual ) {
-		return ObjectUtil.isNullOrEmptyString( expected ) ? !ObjectUtil.isNullOrEmptyString( actual )
-				: !expected.equals( actual );
+		if ( ObjectUtil.isNullOrEmptyString( expected ) ) {
+			return !ObjectUtil.isNullOrEmptyString( actual );
+		} else {
```
Comment (diba1013) [inline src/main/java/de/retest/ui/diff/IdentifyingAttributesDifferenceFinder.java]: This else can be removed.

File: src/main/java/de/retest/ui/diff/IdentifyingAttributesDifferenceFinder.java [hunk 7]
```diff
@@ -67,8 +64,14 @@ private static boolean pathDiffers( final IdentifyingAttributes expected, final
...
+			if ( Objects.isNull( actual ) ) {
+				return false;
+			}
+			return !expected.equals( actual );
+		}
```
Comment (diba1013) [inline src/main/java/de/retest/ui/diff/IdentifyingAttributesDifferenceFinder.java]: I would prefer to use boolean expressions chaining

File: src/main/java/de/retest/ui/diff/IdentifyingAttributesDifferenceFinder.java [hunk 8]
```diff
@@ -67,8 +64,14 @@ private static boolean pathDiffers( final IdentifyingAttributes expected, final
 
 	// We treat null == "". This is OK for visible attributeDifferences but less suitable for invisible ones.
 	private static boolean differs( final Object expected, final Object actual ) {
-		return ObjectUtil.isNullOrEmptyString( expected ) ? !ObjectUtil.isNullOrEmptyString( actual )
-				: !expected.equals( actual );
+		if ( ObjectUtil.isNullOrEmptyString( expected ) ) {
+			return !ObjectUtil.isNullOrEmptyString( actual );
+		} else {
+			if ( Objects.isNull( actual ) ) {
```
Comment (diba1013) [inline src/main/java/de/retest/ui/diff/IdentifyingAttributesDifferenceFinder.java]: Did you fix a bug here since this is not in the old code?

File: src/main/java/de/retest/ui/diff/RootElementDifferenceFinder.java [hunk 9]
```diff
@@ -78,7 +78,7 @@ private RootElement findAndRemove( final RootElement expected, final List<RootEl
 	}
 
 	private double compareContainedComponents( final RootElement expected, final RootElement bestWindow ) {
-		return getNumberOfChildDifferences( expected, bestWindow )
+		return (double) (getNumberOfChildDifferences( expected, bestWindow ))
 				/ (getNumberOfChildComponents( expected, bestWindow ) + 1);
```
Comment (diba1013) [inline src/main/java/de/retest/ui/diff/RootElementDifferenceFinder.java]: You could just juse  `+ 1.0`
```suggestion 
return getNumberOfChildDifferences( expected, bestWindow ) / (getNumberOfChildComponents( expected, bestWindow ) + 1.0);
```

File: src/main/java/de/retest/util/ReflectionUtilities.java [hunk 10]
```diff
@@ -22,7 +22,7 @@
 
 	public static List<Field> getAllFields( Class<?> clazz ) {
 		final List<Field> result = new ArrayList<>();
-		while ( !clazz.getName().equals( "java.lang.Object" ) ) {
+		while ( clazz.isAssignableFrom( Object.class ) ) {
```
Comment (diba1013) [inline src/main/java/de/retest/util/ReflectionUtilities.java]: isAssignableFrom checks for Instance and superclass, thus you would also take into account `Object`

**Review by diba1013 — DISMISSED** [review]

**Review by beatngu13 — CHANGES_REQUESTED** [review]
The changes are sometimes hard to follow, they should have been separated into multiple micro commits with explaining commit messages.

File: src/main/java/de/retest/persistence/migration/XmlTransformer.java [hunk 11]
```diff
@@ -39,37 +38,39 @@
 
 	public final InputStream transform( final InputStream inputStream ) {
 		try {
-			reset();
```
Comment (diba1013) [inline src/main/java/de/retest/persistence/migration/XmlTransformer.java:42]: Must the reset be on top?

**Review by beatngu13 — APPROVED** [review]
